tips/TIP-0038/tip-0038.md

```diff
@@ -0,0 +1,928 @@
+deposit requirements must be covered via IOTA coins.
+
+User-defined tokens are called _Native Tokens_ on the protocol level. The maximum supply of a particular native token is
+defined by the representation chosen on the protocol level for defining their amounts in outputs. Since native tokens
+are also a vehicle to wrap layer 2 tokens into layer 1 tokens, the chosen representation must take into account the
+maximum possible supply of layer 2 tokens. Solidity, the most popular smart contract language defines the maximum supply
+of an ERC-20 token as `MaxUint256`, therefore it should be possible to represent such a huge amount of assets on
+layer 1.
+
+Outputs that can carry Native Tokens have a [_Native Token Feature_](#native-token-feature). Its transaction validation
+rules are defined in that section.
+
+## Chain Constraint in UTXO
+
+Previously created transaction outputs are destroyed when they are consumed in a subsequent transaction as an input. The
+chain constraint makes it possible to **carry the UTXO state machine state encoded in outputs across transactions.**
+When an output with chain constraint is consumed, that transaction has to create a single subsequent output that carries
+the state forward. The **state can be updated according to the transition rules defined for the given type of output and
+its current state**. As a consequence, each such output has a unique successor, and together they form a path or _chain_
+in the graph induced by the UTXO spends. Each chain is identified by its globally unique identifier.
+
+![](chain-constraint.png)
+
+Account outputs, foundry outputs, and NFT outputs all use this chain constraint concept and define their own unique
+identifiers.
+
+### Unlocking Chain Script Locked Outputs
+
+Outputs with a chain constraint receive their unique identifiers upon creation, generated by the protocol, and carry it
+forward with them through transactions until they are destroyed. These unique identifiers also function as global
+addresses for the state machines, but unlike <i>Ed25519 Addresses</i>, they are not backed by private keys that could be
+used for signing. The rightful owners who can unlock these addresses are defined in the outputs themselves.
+
+Since such addresses are accounts in the ledger, it is possible to send funds to these addresses. The unlock mechanism
+of such funds is designed in a way that **proving ownership of the address is reduced to the ability to unlock the
+corresponding output that defines the address.**
+
+## Addresses
+
+The following section shows the basic supported Address Types and their serialization.
+
+| Address | Type Byte as `uint8` | Bech32 Encoded |
+| ------- | -------------------- | -------------- |
+| Ed25519 | 0                    | iota1**q**...  |
+| Account | 8                    | iota1**p**...  |
+| NFT     | 16                   | iota1**z**...  |
```

Comment:
I generally avoid any kind of backlinking references because it's a nightmare to maintain.
If another TIP comes around in the future that adds another address type, this list will once again be outdated and hardly anyone will remember to update it.
So I generally try to write TIPs so that they can remain stable and don't need to be updated / become outdated even when new TIPs are proposed, with the trade-off that we never have a global list of all address types or all features types, ... in one TIP.

Reply:
Not sure if I follow, this table was defined in TIP 18 already
In the TIP it mentions `Many of these were originally introduced in [TIP-18](https://github.com/iotaledger/tips/blob/tip38/tips/TIP-0018/tip-0018.md), and the functionality defined in this document is an extension of the primitives defined in TIP-18.`
Shouldn't there then rather be a table with only the two new addresses?
Also in https://github.com/iotaledger/tips/blob/tip38/tips/TIP-0038/tip-0038.md#address-unlock-condition all addresses are listet

I see the additional part for the table is here for the restricted address https://github.com/iotaledger/tips-draft/blob/tip50/tips/TIP-0050/tip-0050.md#restricted-address

but in https://github.com/iotaledger/tips-draft/blob/tip52/tips/TIP-0052/tip-0052.md#bech32-representation there is nothing (at the bottom an example where one can see it's `y`, but not as this table form)

On a side note, why has TIP 50 the higher (40) type byte than TIP52 (32)? Don't care that much about it, just noticed it now :trollface: 

Don't get me wrong, I don't have a big issue with anything here, just some observation
So feel free to resolve the conversation if you think it's good as it is

Reply:
Well, TIP-18 is completely deprecated for IOTA 2.0 so we have to redefine this stuff, and for the those three address types that happens in this TIP.

Nice observation about the missing bech32 table, I just added it to TIP-52! 

When we started the TIPs my goal was also to have them build on each other successively without and circular references, but that is not possible due to many small details. Some unlock conditions need to reference output types (like Storage Deposit Return UC needs to define that the return amount must be deposited via a Basic Output), so already we have a TIP-38 -> TIP-41 dependency. And there are many more of those, unfortunately.

So that led me to drop the efforts of trying to increase type numbers in line with the TIPs. Also, we try for them to make sense on the type range they're defined on, which is not necessarily the order of the TIPs. One can only optimize for one logical dimension here :smile:.